### ISSUE TITLE
Adding regularisation causes NaNs on first Epoch

### ISSUE THREAD

**Issue by deveshjawla** [FluxML/Flux.jl#1254]
I am training for a classification problem. The input is 518 length normalized vector and the labels are one hot vectors of length 2. The data is Float32. The model trains on CPU, however, fails on CUDA GPU TITAN RTX. If I remove the regularisation, then the model trains on the GPU as well.

What I have tried:
1. Varying batch size
2. varying the learning rate
3. Different optimisers with varying learning rates
4. Different cost functions and corresponding output layer activations
5. Different Architectures of the model i.e. one hidden layer and more than one

The important pieces of code are as follows.

```
function build_model(args, ; n_features=518, nclasses=2)
    return Chain(Dense(n_features, 16, relu),Dropout(0.25),Dense(16,8,relu),Dropout(0.5),Dense(8,4,relu),Dense(4, nclasses), softmax)
end

loss(x,y) = crossentropy(model(x), y) + sum(norm,params(model))

opt = AdaMax(args.η)

Flux.train!(loss, params(model), train_set, opt )
```
[Link to source code](https://github.com/deveshjawla/Masters_cluster_codes/blob/master/deeplearning_gonihedric/bin_code.jl)

**Comment by arnold-dev** [comment]
Have you tried to reduce the regularization importance (with lambda parameter below)? Adding an epsilon to crossentropy can help too.

```julia
const ε = 1.0f-10
λ = 0.001 # weight decay
loss(x, y) = crossentropy(model(x) .+ ε, y .+ ε) + λ * sum(norm, params(model))
```

**Comment by CarloLucibello** [comment]
L2 regularization is typically implemented using 
```julia
L2reg(m) = sum(sum(p.^2) for p in Flux.params(m))
```
or similar alternatives. With `norm` you also take the sqrt of the weights' sum, which is usually not done in practice and may lead to instabilities when computing gradients. We should fix this in the docs. 

Also, for the sake of numerical stability, I suggest using `logitcrossentropy` instead of `crossentropy` and remove the last `softmax` operation in the model

**Comment by deveshjawla** [comment]
> Have you tried to reduce the regularization importance (with lambda parameter below)? Adding an epsilon to crossentropy can help too.
> 
> ```julia
> const ε = 1.0f-10
> λ = 0.001 # weight decay
> loss(x, y) = crossentropy(model(x) .+ ε, y .+ ε) + λ * sum(norm, params(model))
> ```

Yes, I tried with smaller lambda but it didn't help.
I will try the epsilon, it could be that outputs are going to zero. I don't know how to check when the numbers go to zero.

**Comment by deveshjawla** [comment]
> L2 regularization is typically implemented using
> 
> ```julia
> L2reg(m) = sum(sum(p.^2) for p in Flux.params(m))
> ```
> 
> or similar alternatives. With `norm` you also take the sqrt of the weights' sum, which is usually not done in practice and may lead to instabilities when computing gradients. We should fix this in the docs.
> 
> Also, for the sake of numerical stability, I suggest using `logitcrossentropy` instead of `crossentropy` and remove the last `softmax` operation in the model

Yes, thanks. That makes sense, I just went along with the documentation. It's true there is no sqrt on the L2 term.
However, when iterating through the params(m), there's also bias vectors. Which we don't take in the L2. How would you explain this?

Yes, logitcrossentropy is stable as per docs. However the problem persisted. I removed the softmax and use logitCE. Because the issue was with the regularisation term. If I remove it, then everything worked fine.

**Comment by arnold-dev** [comment]
Maybe this is a solution to your problem: #939

**Comment by CarloLucibello** [comment]
So adding proper L2 reg instead of norm should solve your problem. In fact, the problem is that your bias is 0 at init, and the gradient of  `norm(p)` gives factor `1/norm(p)` resulting in NaNs 

> However, when iterating through the params(m), there's also bias vectors. Which we don't take in the L2. How would you explain this?

I think one commonly regularizes biases as well. If you want to do something different look at @arnold-dev pointer to filter params

**Comment by deveshjawla** [comment]
> Maybe this is a solution to your problem: #939

Yes one can always filter out the bias vectors.

**Comment by deveshjawla** [comment]
> So adding proper L2 reg instead of norm should solve your problem. In fact, the problem is that your bias is 0 at init, and the gradient of `norm(p)` gives factor `1/norm(p)` resulting in NaNs
> 
> > However, when iterating through the params(m), there's also bias vectors. Which we don't take in the L2. How would you explain this?
> 
> I think one commonly regularizes biases as well. If you want to do something different look at @arnold-dev pointer to filter params

Thanks. that does it.))
